### ISSUE TITLE
gitian: GLIBC_2.29 not found error at bitcoind startup on bionic

### ISSUE THREAD

**Issue by willyko** [bitcoin/bitcoin#21454]
bitcoind & bitcoin-qt built from gitian on master branch have not been working on Bionic for the last month or so.  
The bitcoind works on Focal though.
Following @jonasschnelli 's nightly gitian build, the last working bitcoind binary was dated January 28 on commit d1ddead0

**Expected behavior**
bitcoind to run 

**Actual behavior**
```
$ ./bitcoind
bitcoind: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by bitcoind)
bitcoind: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by bitcoind)
```

**To reproduce**
On Ubuntu Bionic, download any of the bitcoin-*-x86_64-linux-gnu.tar.gz after January 28th and run the bitcoind

**System information**
The binaries are downloaded from https://bitcoin.jonasschnelli.ch/gitian/?show=nightly#nighly
I ran the gitian build myself and hit the same error too.
The binaries work on Focal but not on Bionic.  Did not test on other Ubuntu releases

**Comment by MarcoFalke** [comment]
6a726cb534ed471010bf3f466732874b2a1b22e5 is from the 28th

**Comment by MarcoFalke** [comment]
guix is also affected

**Comment by hebasto** [comment]
FWIW,
```
$ lsb_release -d
Description:	Ubuntu 18.04.5 LTS
$ ldd --version | head -1
ldd (Ubuntu GLIBC 2.27-3ubuntu1.4) 2.27
```

**Comment by hebasto** [comment]
> [6a726cb](https://github.com/bitcoin/bitcoin/commit/6a726cb534ed471010bf3f466732874b2a1b22e5) is from the 28th

Just tested the gitian build of 6a726cb534ed471010bf3f466732874b2a1b22e5 on bionic.
It works flawlessly.

**Comment by sipa** [comment]
@jonasschnelli How are your nightlies built? Do they use gitian?

**Comment by willyko** [comment]
There were no nightly build from January 29 - February 10, and it looks like some builds were failing due to ` No space left on device` error according to the log.  I was able to build and run on bionic all the way up to https://github.com/bitcoin/bitcoin/commit/c969ab43c3888b1cd8e691484f2dc32a607d050f and that's the commit right before the switch to focal for gitian vm

**Comment by hebasto** [comment]
I can confirm that issue is caused by ca85449f22cb909b9ca83861aa51fbf8f40a9070 (#21036 merging) -- gitian builds work fine before this commit.

**Comment by fanquake** [comment]
The offending symbols  are:
```bash
0000000000000000      DF *UND*	0000000000000000  GLIBC_2.29  log
0000000000000000      DF *UND*	0000000000000000  GLIBC_2.29  pow
0000000000000000      DF *UND*	0000000000000000  GLIBC_2.29  exp
0000000000000000      DF *UND*	0000000000000000  GLIBC_2.28  fcntl64
```

**Comment by hebasto** [comment]
> The offending symbols are:

Should they be wrapped in `glibc_compat.cpp`?

**Comment by fanquake** [comment]
I was confused as to how this could have happened, because our gitian/guix builds (symbol checks) all should have failed in #21036 if there were newer-than-allowed symbols being introduced. The issue seems to be that [Pixie](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/pixie.py) is incorrectly returning `None`, rather than a symbol version quite often. i.e with:
```diff
index 436f179d6..831346c5b 100755
--- a/contrib/devtools/symbol-check.py
+++ b/contrib/devtools/symbol-check.py
@@ -174,6 +174,7 @@ def check_imported_symbols(filename) -> bool:
             continue
         sym = symbol.name.decode()
         version = symbol.version.decode() if symbol.version is not None else None
+        print(sym, version)
         if version and not check_version(MAX_VERSIONS, version, elf.hdr.e_machine):
             print('{}: symbol {} from unsupported version {}'.format(filename, cppfilt(sym), version))
             ok = False
```
I'm seeing this:
```bash
<trim>
fcntl64 None
memmove None
pthread_rwlock_unlock None
realloc None
__fprintf_chk None
fseeko64 None
pthread_once None
_Unwind_GetRegionStart None
__pthread_key_create None
getgrgid None
fchown None
__tls_get_addr GLIBC_2.3
__memcpy_chk None
_Exit None
open64 None
freeifaddrs None
pow None
_ITM_deregisterTMCloneTable None
strpbrk None
pthread_getschedparam None
_Unwind_Resume None
timerfd_create None
getauxval None
getprotobynumber None
strcat None
mmap None
umask None
setsockopt None
getpgid None
exp None
_Unwind_SetIP None
<trim>
```

when running `python3 bitcoin/contrib/devtools/symbol-check.py bitcoin-ee0dc02c6f93/bin/bitcoind`. Which is against a recent nightly build downloaded from bitcoin builds.

**Comment by sidhujag** [comment]
> The offending symbols are:
> 
> ```shell
> 0000000000000000      DF *UND*	0000000000000000  GLIBC_2.29  log
> 0000000000000000      DF *UND*	0000000000000000  GLIBC_2.29  pow
> 0000000000000000      DF *UND*	0000000000000000  GLIBC_2.29  exp
> 0000000000000000      DF *UND*	0000000000000000  GLIBC_2.28  fcntl64
> ```

Besides the point but pow on <= 2.28 is 10k times slower if you don't need double precision you can use powf(), I found this seemed to fix the unresolved export for focal in our code. pow() is used in traffic widget and addrman (GetChance) so not particularly sensitive imo.

**Comment by laanwj** [comment]
> The issue seems to be that Pixie is incorrectly returning None, rather than a symbol version quite often. i.e with:

Interesting. Will take a look at the pixie issue.

**Comment by hebasto** [comment]
`bitcoind` compatibility fixed in #22281 and #22287.

@willyko Do you mind testing suggested fixes?

**Comment by hebasto** [comment]
Since #22365 has been merged, this could be closed, no?
